File: app/javascript/channels/video_call_channel.js
// app/javascript/channels/video_call_channel.js
import consumer from "./consumer"

const videoCallChannel = consumer.subscriptions.create({ channel: "VideoCallChannel", room: "default_room" }, {
  connected() {
    console.log("Connected to the VideoCallChannel")
  },

  disconnected() {
    console.log("Disconnected from the VideoCallChannel")
  },

  received(data) {
    switch(data.type) {
      case 'offer':
        handleOffer(data.data)
        break;
      case 'answer':
        handleAnswer(data.data)
        break;
      case 'ice_candidate':
        handleIceCandidate(data.data)
        break;
    }
  },

  sendOffer(offer) {
    this.perform('send_offer', { offer: offer })
  },

  sendAnswer(answer) {
    this.perform('send_answer', { answer: answer })
  },

  sendIceCandidate(candidate) {
    this.perform('send_ice_candidate', { candidate: candidate })
  }
})

export default videoCallChannel

File: app/channels/video_call_channel.rb
# app/channels/video_call_channel.rb
class VideoCallChannel < ApplicationCable::Channel
  def subscribed
    stream_from "video_call_#{params[:room]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_offer(data)
    ActionCable.server.broadcast("video_call_#{params[:room]}", { type: 'offer', data: data['offer'] })
  end

  def send_answer(data)
    ActionCable.server.broadcast("video_call_#{params[:room]}", { type: 'answer', data: data['answer'] })
  end

  def send_ice_candidate(data)
    ActionCable.server.broadcast("video_call_#{params[:room]}", { type: 'ice_candidate', data: data['candidate'] })
  end
end
